api/system-apps/update

```diff
@@ -58,6 +62,119 @@ sub has_shortcut
+{
+    my $app = shift;
+    my $db = shift;
+    my $pins = $db->get_prop('cockpit.socket', 'Pins') || '';
+    foreach (split(",",$pins)) {
+        return 1 if ($_ eq $app);
+    }
+    return 0;
+}
+
+sub update_pins_file
+{
+    my @pinned_apps;
+    my @pinned_apps_names;
+    my %shortcuts;
+    my %pins;
+    my $pinned_apps_file;
+
+    if ($< == 0) { # non-root users can't access the config db
+        my $db = esmith::ConfigDB->open_ro();
+
+        my $shortcuts = $db->get_prop('cockpit.socket', 'Shortcuts') || '';
+        foreach (split(",",$shortcuts)) {
+            $shortcuts{$_} = 1;
+        }
+
+        my $pins = $db->get_prop('cockpit.socket', 'Pins') || '';
+        foreach (split(",",$pins)) {
+            $pins{$_} = 1;
+            push(@pinned_apps_names, $_);
+        }
+    }
+
+    foreach my $f (glob("$path*.json")) {
+        my ($name,$path,$suffix) = fileparse($f,(".json"));
+        if (grep(/^$name$/, @pinned_apps_names)) {
+            my $data = app_info($name, \%shortcuts, \%pins) || next;
+            push(@pinned_apps, $data);
+        }
+    }
+
+    # Check that pinned apps directory exists
+    my ( $file, $directories ) = fileparse $pinned_apps_file_path;
+
+    if ( !-d $directories ) {
+        make_path $directories or die "Failed to create path: $directories";
+    }
+
+    # Create the file
+    unless(open $pinned_apps_file, '>'.$pinned_apps_file_path) {
+        # Die with error message if we can't open it
+        die "\nUnable to create $pinned_apps_file_path\n";
+    }
+    print $pinned_apps_file encode_json(\@pinned_apps);
+    close $pinned_apps_file;
+}
+
+sub app_info
+{
+    my $app = shift;
+    my $shortcuts = shift;
+    my $pins = shift;
+    my $file = "$path$app.json";
+    my $data = read_json($file);
+    my $extended_data = {};
+    my $encoded;
+    my $raw_string;
+
+    # If available, invoke the infoapi provider
+    # to extend the manifest data dynamically
+    if($data->{'infoapi'}) {
+        if(-x '/usr/libexec/nethserver/api/' . $data->{'infoapi'}->{'path'}) {
+            $extended_data = invoke_info_api(
+                '/usr/libexec/nethserver/api/' . $data->{'infoapi'}->{'path'},
+                safe_decode_json($data->{'infoapi'}->{'input'})
+            );
+        } else {
+            warn "[ERROR] missing executable script! Actual `infoapi.path` value: " . $data->{'infoapi'}->{'path'} . "\n";
+        }
+    }
+    $data = {%$data, %$extended_data}; # merges the input hashes
+
+    # Set hardcoded attributes:
+    $data->{'shortcut'} = defined($shortcuts->{$app}) ? 1 : 0;
+    $data->{'pin'} = defined($pins->{$app}) ? 1 : 0;
+    $data->{'editable'} = 1;
+    # Base64 icon
+    open (my $image, "$cockpit_path$app/" . $data->{'icon'}) or die "$!";
+    $raw_string = do{ local $/ = undef; <$image>; };
+    $encoded = encode_base64($raw_string);
+    $encoded =~ s/\R//g;
+    $data->{'iconBase64'} = $encoded;
+    return $data;
+}
+
+sub invoke_info_api
```

Comment:
See if this function is worth to move to NethServer::ApiTools (`/usr/share/perl5/vendor_perl/NethServer/ApiTools.pm`)